### PR TITLE
Remove Image Texture requirement for some shaders

### PR DESCRIPTION
**Pull request by loganmc10**
This removes the Image Texture requirement for the fog effect in Beetle Adventure Racing and the depth in Resident Evil 2.

Tested on Windows + Intel. I haven't tested this on GLES.

@fzurita This should make RE2 work in GLES I would assume. It should also make the BAR fog work on all devices. Hopefully it might fix the bug in https://github.com/gonetz/GLideN64/issues/1777

### CODE REVIEW

**Comment by loganmc10** [comment]
Ok I disabled it for GLES2, GLES2 doesn't support integer textures either, so the shader would need to be modified as well as a few other things.

**Comment by fzurita** [comment]
I just tried it. RE2 still does work, at least with Adreno. It doesn't fix BAR menus with Adreno either.

**Comment by loganmc10** [comment]
Did you have fragment depth writing enabled for RE2? That is very strange.

The BAR thing is really bizarre. Either way I think this is a good change since it removes the Image Texture requirement, it's too bad that bug still exists though.

**Comment by fzurita** [comment]
Yes, I had fragment depth write enabled.

**Comment by loganmc10** [comment]
Well the good news is now that it doesn't depend on Image Textures, I can test this GLES code on my laptop (supports GLES3.0). I pushed a change that may fix the BAR menus, they were broken on my laptop in GLES mode (the usampler2D needed a precision set).

I notice that RE2 has clipping disabled, and it is also broken on my laptop in GLES mode. I assume it might need the recent depth clamp emulation to work (I'll have to test on my Adreno device that supports that soon)

**Comment by fzurita** [comment]
The menu still doesn't work with Adreno, I think that's just a driver bug.

**Comment by fzurita** [comment]
Why did you switch the precision from highp to lowp for zlut and tlut images?

**Comment by loganmc10** [comment]
lowp is the default precision qualifier for a sampler2D. usampler2D has no default, so I assumed lowp would also be a good default. My testing showed no difference. I'm not actually sure what the difference between lowp and highp is when you're dealing with integers. But everything I read says lowp is better for performance.

**Comment by fzurita** [comment]
My assumption would be something like 8/16/32 bit. Higher bits allow for bigger numbers. But you are right, normally the scale here is 0 to 255, so lowp would be fine.

**Comment by loganmc10** [comment]
Well that uniform is actually just holding the name of the texture (like the ID number), it's not storing the texture data, so its just the GL identifier for the texture. I don't think high precision would do anything differently.

**Comment by fzurita** [comment]
Yeah, another good point.

**Comment by loganmc10** [comment]
Anyway we now know that the BAR menu works with Mesa GLES, Mali, Nvidia, and other desktop drivers. That combined with the weird glBlit behaviour makes it pretty conclusive that there is some really weird Adreno bug with the BAR menu.

**Comment by fzurita** [comment]
Yep, I completely agree. I have an idea of a workaround that is efficient as well, I just need to find the time to implement it.

**Comment by loganmc10** [comment]
I pushed one unrelated commit since it will fix a regression caused by https://github.com/gonetz/GLideN64/commit/af6460ff4f40f0d0145afcd80ae8707d9dd33b6f

Since that commit only fixes clipping on GLES when GL_NV_shader_noperspective_interpolation is supported, the clipping hack in GraphicsDrawer needs to be enabled always for GLESX.

I also changed the check for if FragmentDepthWrite is supported to if it is enabled. We shouldn't really care if it is supported, it can be enabled or disabled, and we force it to be disabled for GLES2 anyway

File: src/Graphics/Context.h [hunk 1]
```diff
@@ -17,12 +17,13 @@ namespace graphics {
...
 	enum class SpecialFeatures {
 		Multisampling,
-		FragmentDepthWrite,
 		BlitFramebuffer,
 		WeakBlitFramebuffer,
 		DepthFramebufferTextures,
 		ShaderProgramBinary,
-		ImageTextures
+		ImageTextures,
+		LUTTextures,
```
Comment (gonetz) [inline src/Graphics/Context.h]: LUTTextures is a bad name. Can you pick more general name, which reflect essence of that special feature?

File: src/Graphics/Context.cpp [hunk 2]
```diff
@@ -20,8 +17,6 @@ void Context::init()
 	m_impl.reset(new opengl::ContextImpl);
 	m_impl->init();
 	m_fbTexFormats.reset(m_impl->getFramebufferTextureFormats());
-	imageTextures = isSupported(SpecialFeatures::ImageTextures);
-	multisampling = isSupported(SpecialFeatures::Multisampling);
```
Comment (gonetz) [inline src/Graphics/Context.cpp]: Why? These variables cached most often requested special features.

Comment (gonetz) [reply src/Graphics/Context.cpp]: You may add variable for LUTTexture as well.

File: src/Graphics/OpenGLContext/opengl_TextureManipulationObjectFactory.cpp [hunk 3]
```diff
@@ -191,7 +191,6 @@ namespace opengl {
 				}
 
 				if (_params.ImageUnit.isValid()) {
-					assert(IS_GL_FUNCTION_VALID(glBindImageTexture));
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/opengl_TextureManipulationObjectFactory.cpp]: Why removed?

File: src/Graphics/OpenGLContext/opengl_TextureManipulationObjectFactory.cpp [hunk 4]
```diff
@@ -273,7 +272,6 @@ namespace opengl {
 				_params.data);
 
 			if (_params.ImageUnit.isValid() && _params.internalFormat.isValid()) {
-				assert(IS_GL_FUNCTION_VALID(glBindImageTexture));
```
Comment (gonetz) [inline src/Graphics/OpenGLContext/opengl_TextureManipulationObjectFactory.cpp]: Why removed?

**Comment by gonetz** [comment]
> Ok I changed all the Special Features checks to use Context:: style checks, this should be faster and keep everything consistent. Is that alright?

Yes, looks good. Thanks!

**Pull request merged**